File: workers/ruby/apps/registry.rb
# frozen_string_literal: true

require 'harness'
require 'optparse'
require_relative 'worker/app'

module Apps
  module Registry
    DEFAULT_APP_NAME = 'worker'
    APPS = {
      'worker' => Worker.method(:app)
    }.freeze

    module_function

    def run(argv = ARGV)
      argv = Array(argv).dup
      app_name = DEFAULT_APP_NAME
      OptionParser.new do |opts|
        opts.on('--app APP') { |value| app_name = value }
      end.order!(argv)

      app = APPS[app_name]
      raise SystemExit, "unknown Ruby worker app #{app_name.inspect}" if app_name.empty? || app.nil?

      Harness.run(app.call, argv)
    end
  end
end

File: workers/python/apps/registry.py
from __future__ import annotations

import argparse
import sys
from collections.abc import Sequence

from apps.helloworld.app import app as helloworld_app
from apps.worker.app import app as worker_app
from harness import App, run

DEFAULT_APP_NAME = "worker"

registry: dict[str, App] = {
    DEFAULT_APP_NAME: worker_app,
    "helloworld": helloworld_app,
}


def main(argv: Sequence[str] | None = None) -> None:
    parser = argparse.ArgumentParser(add_help=False, allow_abbrev=False)
    parser.add_argument("--app", default=DEFAULT_APP_NAME)
    args, remaining = parser.parse_known_args(sys.argv[1:] if argv is None else argv)

    app = registry.get(args.app)
    if app is None:
        raise SystemExit(f"unknown Python worker app {args.app!r}")

    run(app, remaining)


if __name__ == "__main__":
    main()
